Return error if result of a txn commit is unknown

spencerkimball opened the pull request:
Introduce `AmbiguousCommitError` in the event that a batch with an
`EndTransaction` request is sent but the response is not available.

Fixes #6053, #7604, and #10023

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10154)

<!-- Reviewable:end -->

spencerkimball committed:
Return error if result of a txn commit is unknown

Introduce `AmbiguousCommitError` in the event that a batch with an
`EndTransaction` request is sent but the response is not available.

Fixes #6053, #7604, and #10023

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> The move-leaseholder-to-front is more worrisome than the ambiguous commit change. 

---

Review status: 0 of 14 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-4e_6qsQPbrhhmpk:-KUm-4e_6qsQPbrhhmpl:b-n0fivq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L963)):*
> ```Go
> 	var haveCommit bool
> 	if etArg, ok := args.GetArg(roachpb.EndTransaction); ok &&
> 		etArg.(*roachpb.EndTransactionRequest).Commit {
> ```

Is there any danger to retrying ambiguous aborts? Might be worth a comment explaining why.

---

*[pkg/kv/dist_sender.go, line 1039 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-e1xj-ARnzE2XLjJ:-KUm-e1xj-ARnzE2XLjK:b8sn6qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L1039)):*
> ```Go
> 					log.Warningf(opts.ctx, "RPC error: %s", err)
> 				}
> 				// All connection errors except for an unavailable node (this
> ```

What about our circuit breaker errors?

---

*[pkg/kv/dist_sender.go, line 1090 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm0e6WFM6M0WdRFVfE:-KUm0e6WFM6M0WdRFVfF:b38m7y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L1090)):*
> ```Go
> 			// Move the new lease holder to the head of the queue for the next retry.
> 			if err := transport.MoveToFront(leaseHolder); err != nil {
> 				log.Warningf(ds.Ctx, "unable to move leaseholder %+v to front of transport", leaseHolder)
> ```

Include `err` in the log message?

---

*[pkg/kv/transport.go, line 219 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm3NXYgbf1Cb7t-LFs:-KUm3NXYgbf1Cb7t-LFt:b-tpohye) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/transport.go#L219)):*
> ```Go
> // clients. If found and not currently pending, it is moved to the front,
> // even if it's already been invoked.
> func (gt *grpcTransport) MoveToFront(replica roachpb.ReplicaDescriptor) error {
> ```

Can this be unittested? The interaction between failures and `MoveToFront` seems complicated. I wonder if we can get into bad retry loops where we send an RPC to a replica, it responds that it isn't the lease holder and doesn't know who is. We talk to another replica and it says the first replica is the lease holder. Because of the way `MoveToFront` adjusts `clientIndex` I can see scenarios where we retry excessively.

---

*[pkg/kv/transport.go, line 240 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1LHvuOdhBOeiTqs_:-KUm1LHwA4tJqNMl0Uqm:b-7gx93q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/transport.go#L240)):*
> ```Go
> 		}
> 	}
> 	return errors.Errorf("replica %+v not part of GRPC transport", replica)
> ```

There isn't much utility in returning this error. 

---

*[pkg/sql/ambiguous_commit_test.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1YpJ5YUN64zgG-KT:-KUm1YpJ5YUN64zgG-KU:b4vw31w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/sql/ambiguous_commit_test.go#L48)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	// Create a command filter which prevents EndTransaction from succeeding.
> ```

`s/from succeeding/from returning a response/g`

---

*[pkg/storage/replica_command.go, line 2842 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1vaM5bO5C4UKHgoe:-KUm1vaM5bO5C4UKHgof:btaizom) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/storage/replica_command.go#L2842)):*
> ```Go
> 			// Commit this batch on its own to ensure that the transaction record
> 			// is created in the right place (our triggers rely on this).
> 			log.Warningf(ctx, "updating LHS descriptor")
> ```

Why is this a warning? Revert?

---

*[pkg/storage/storagebase/base.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm25sjVWQYK2JvMYmJ:-KUm25sjVWQYK2JvMYmK:buobrtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/storage/storagebase/base.go#L52)):*
> ```Go
> // ReplicaResponseFilter is used in unittests to modify the outbound
> // response returned to a waiting client after a replica command has
> // been processed. This filter is used only by the command proposer.
> ```

`s/used only/invoked only/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-:-KUm31PNiHxw5vj7-IxR:b8365nh)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 14 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-4e_6qsQPbrhhmpk:-KUmQKehCga6kIo6qvhZ:b-l73tkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L963)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is there any danger to retrying ambiguous aborts? Might be worth a comment explaining why.
</details>
There is no danger. Here's comment I'm adding:

```
// We retry ambiguous commit batches in order to achieve success before resorting
// to returning the unrecoverable AmbiguousCommitError. This is safe because
// repeating an already-successfully applied batch is guaranteed to return either a
// TransactionReplayError (in case the replay happens at the original leader), or a
// TransactionRetryError (in case the replay happens at a new leader). If the original
// attempt merely timed out or was lost, then the batch will succeed and we can be
// assured the commit was applied just once. 
```

---

*[pkg/kv/dist_sender.go, line 1039 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-e1xj-ARnzE2XLjJ:-KUmTyGvvCtMKfORLwoD:bdxjlfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L1039)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What about our circuit breaker errors?
</details>
I couldn't for the life of me find out where these come into play in our use of GRPC. I concluded that we're not using circuit breakers for this purpose in the code and are instead relying on GRPC's fail fast determination. Seems we use them for gossip clients and for raft transport but not GRPC.

---

*[pkg/kv/dist_sender.go, line 1090 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm0e6WFM6M0WdRFVfE:-KUmUQFmSUKawJyLNm0k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L1090)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Include `err` in the log message?
</details>
Done.

---

*[pkg/kv/transport.go, line 219 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm3NXYgbf1Cb7t-LFs:-KUmUWC9B5QwWhJYBJ3P:b-bdljk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/transport.go#L219)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Can this be unittested? The interaction between failures and `MoveToFront` seems complicated. I wonder if we can get into bad retry loops where we send an RPC to a replica, it responds that it isn't the lease holder and doesn't know who is. We talk to another replica and it says the first replica is the lease holder. Because of the way `MoveToFront` adjusts `clientIndex` I can see scenarios where we retry excessively.
</details>
Hmm, also if we had five replicas and two of the five were partitioned and each thought the other was the leader. We could ping back and forth. I've changed it to limit each client to a single retry.

Added a unittest as well.

---

*[pkg/kv/transport.go, line 240 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1LHvuOdhBOeiTqs_:-KUmcOpOymojEM2sLJ6Y:bs36ml6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/transport.go#L240)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> There isn't much utility in returning this error. 
</details>
OK, will just ignore if replica isn't present.

---

*[pkg/sql/ambiguous_commit_test.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1YpJ5YUN64zgG-KT:-KUmeu9ogwghzbiZeqS2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/sql/ambiguous_commit_test.go#L48)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/from succeeding/from returning a response/g`
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2842 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1vaM5bO5C4UKHgoe:-KUmf-kJ5PwRO0pIY51R:b43xqq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/storage/replica_command.go#L2842)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Why is this a warning? Revert?
</details>
Oops, reverted.

---

*[pkg/storage/storagebase/base.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm25sjVWQYK2JvMYmJ:-KUmf3fvrvKotC1JH96g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/storage/storagebase/base.go#L52)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/used only/invoked only/g`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-4e_6qsQPbrhhmpk:-KUo1Tol6nV553Ohiql6:b-fuwxwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L963)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> There is no danger. Here's comment I'm adding:
> 
> ```
> // We retry ambiguous commit batches in order to achieve success before resorting
> // to returning the unrecoverable AmbiguousCommitError. This is safe because
> // repeating an already-successfully applied batch is guaranteed to return either a
> // TransactionReplayError (in case the replay happens at the original leader), or a
> // TransactionRetryError (in case the replay happens at a new leader). If the original
> // attempt merely timed out or was lost, then the batch will succeed and we can be
> // assured the commit was applied just once. 
> ```
</details>
Nice comment, but I don't think it addresses my question. You're only setting `ambiguousCommit` if the `EndTransaction` is a "commit", not an "abort". I was wondering if there is any problem with retrying ambiguous aborts. 

---

*[pkg/kv/dist_sender.go, line 1039 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-e1xj-ARnzE2XLjJ:-KUo2LCV0aHetGRAxtCi:bdjw6tt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L1039)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> I couldn't for the life of me find out where these come into play in our use of GRPC. I concluded that we're not using circuit breakers for this purpose in the code and are instead relying on GRPC's fail fast determination. Seems we use them for gossip clients and for raft transport but not GRPC.
</details>
Yeah, we don't seem to be using circuit breakers here. Should we? If yes, let's file an issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-4e_6qsQPbrhhmpk:-KUqdMTX_OnF9QHzZcWO:b-4l8fsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L963)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Nice comment, but I don't think it addresses my question. You're only setting `ambiguousCommit` if the `EndTransaction` is a "commit", not an "abort". I was wondering if there is any problem with retrying ambiguous aborts. 
</details>
I thought you just mis-typed abort in your originally comment. There's no danger on abort because abort only works if there's a txn record present. If it runs twice, we'll get a `TransactionStatusError("does not exist")`, which will presumably look weird at the sql client call site, but isn't harmful. I added another comment.

---

*[pkg/kv/dist_sender.go, line 1039 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm-e1xj-ARnzE2XLjJ:-KUqe0Sjz_K9YKzK8-le:b-e2hscl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/kv/dist_sender.go#L1039)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yeah, we don't seem to be using circuit breakers here. Should we? If yes, let's file an issue.
</details>
We might need to if there's tons of log spam coming out of the distributed senders, but I believe our check for healthy flags and rebalancing of leader leases obviate the need to employ a circuit breaker. Until we see an actual problem, let's not create make work issues.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 14 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 1055 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs5Igkdx_36j1BduXl:-KUs5Igkdx_36j1BduXm:b-wodkzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/dist_sender.go#L1055)):*
> ```Go
> 				// lost, then the batch will succeed and we can be assured the
> 				// commit was applied just once.
> 				if haveCommit && grpc.Code(err) != codes.Unavailable {
> ```

where did you read that this is the error code to check for? aren't there other codes which correspond to the definitely-not-sent condition?

---

*[pkg/kv/dist_sender.go, line 1074 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs7jmmiQcadctcl0hV:-KUs7jmnbEOZkMEyW6Bx:b-di762) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/dist_sender.go#L1074)):*
> ```Go
> 					)
> 				}
> 				log.VEventf(opts.ctx, 2, "%v", err)
> ```

shouldn't this be `log.ErrEvent(opts.ctx, err.Error())`?

---

*[pkg/kv/transport.go, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8D2_tPEtJKcyQm3s:-KUs8D2_tPEtJKcyQm3t:bfjh27b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L102)):*
> ```Go
> 	// front of the ordering of replicas to try. If the replica has
> 	// already been tried, it will be retried. If the specified replica
> 	// can't be found, returns an error.
> ```

It does not appear to return an error.

---

*[pkg/kv/transport.go, line 186 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8TCo9izwvYCwXaRM:-KUs8TCo9izwvYCwXaRN:bjergry) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L186)):*
> ```Go
> 		}
> 
> 		reply, err := localServer.Batch(gt.opts.ctx, &client.args)
> ```

this is local, but can still take appreciable time. is it wise to hold the mutex across this call?

---

*[pkg/kv/transport.go, line 212 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8ZwT3Yzx9CTQLa6n:-KUs8ZwT3Yzx9CTQLa6o:bm7i64m) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L212)):*
> ```Go
> 		defer gt.Unlock()
> 		client.pending = false
> 		done <- BatchCall{Reply: reply, Err: err}
> ```

Seems unwise to hold the lock across this line.

---

*[pkg/kv/transport.go, line 216 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8c3uouW21NTPWoce:-KUs8c3uouW21NTPWocf:b6qzdqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L216)):*
> ```Go
> }
> 
> // MoveToFront locates the specified replica in the slice of transport
> ```

This is repeating the comment on the interface; I suggest removing it altogether.

---

*[pkg/kv/transport.go, line 222 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8nTbH1pbdbeYTwyx:-KUs8nTbH1pbdbeYTwyy:bkw2aua) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L222)):*
> ```Go
> 	gt.Lock()
> 	defer gt.Unlock()
> 	for i, client := range gt.orderedClients {
> ```

probably best to replace this with:
```
for i := range gt.orderedClients {
  client := &gt.orderedClients[i]
```

to avoid the subtlety currently present below.

---

*[pkg/kv/transport_test.go, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9TxbENa__L9WwmLg:-KUs9TxbENa__L9WwmLh:bjx1z3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport_test.go#L41)):*
> ```Go
> 		for i, bc := range gt.orderedClients {
> 			if bc.args.Replica != replicas[i] {
> 				t.Fatalf("expected order %+v; got mismatch at index %d: %+v", replicas, i, bc.args.Replica)
> ```

As discussed in many recent PRs, this is bad practice that results in _this_ line number being reported in failures, which makes it impossible to determine which invocation of this function resulted in an unexpected result. Helpers such as this one should return errors for this reason.

---

*[pkg/roachpb/errors.proto, line 179 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs86S2zyxrlM2gULQT:-KUs86S2zyxrlM2gULQU:b36tqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/roachpb/errors.proto#L179)):*
> ```protobuf
> }
> 
> // An AmbiguousCommitError is true to indicate that an EndTransaction
> ```

"is true"?

---

*[pkg/sql/ambiguous_commit_test.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9uwb8isSX9R7qico:-KUs9uwb8isSX9R7qicp:bsyerkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L80)):*
> ```Go
> 	defer tc.Stopper().Stop()
> 	if err := tc.WaitForFullReplication(); err != nil {
> 		t.Error(err)
> ```

Fatal?

---

*[pkg/sql/ambiguous_commit_test.go, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9s7OfTenR3pRiDG5:-KUs9s7OfTenR3pRiDG6:bkvj6rh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L85)):*
> ```Go
> 	sqlDB := sqlutils.MakeSQLRunner(t, tc.Conns[0])
> 
> 	_ = sqlDB.Exec(`CREATE DATABASE test`)
> ```

please check these errors.

---

*[pkg/sql/ambiguous_commit_test.go, line 121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9px746AXN4EqGRIY:-KUs9px746AXN4EqGRIZ:bstc494) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L121)):*
> ```Go
> 		// Use a connection other than through the node which is the current
> 		// leaseholder to ensure that we use GRPC instead of the local server.
> 		_, err := tc.Conns[leaseHolder.NodeID%3].Exec(`INSERT INTO test.t (v) VALUES (1)`)
> ```

what is this magic `3`?

---

*[pkg/sql/ambiguous_commit_test.go, line 131 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUsA9p_ZtV5ZbXfltQw:-KUsA9p_ZtV5ZbXfltQx:bkjothf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L131)):*
> ```Go
> 	for i, s := range tc.Servers {
> 		if leaseHolder.StoreID != s.GetFirstStoreID() {
> 			err = tc.TransferRangeLease(&tableRangeDesc, tc.Target(i))
> ```

nit: use `:=` to avoid overwriting random errors at the outer scope

---

*[pkg/sql/ambiguous_commit_test.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUsACvHBYMlSpaWmxIz:-KUsACvHBYMlSpaWmxJ-:bkjothf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L140)):*
> ```Go
> 
> 	// Close the wait channel and wait for the error from the pending SQL insert.
> 	if err = <-sqlErrCh; !testutils.IsError(err, "transaction commit result is ambiguous") {
> ```

nit: use `:=` to avoid overwriting random errors at the outer scope

---

*[pkg/sql/ambiguous_commit_test.go, line 146 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUsA5AkiZBJJHPPm4dq:-KUsA5AkiZBJJHPPm4dr:b-1bf5n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L146)):*
> ```Go
> 	// Verify a single row exists in the table.
> 	var rowCount int
> 	sqlDB.QueryRow(`SELECT COUNT(*) FROM test.t`).Scan(&rowCount)
> ```

unchecked error? how did this pass lint?

---

*[pkg/storage/replica_command.go, line 2842 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1vaM5bO5C4UKHgoe:-KUs4_NKKiMYF0jbDvdI:b324822) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/storage/replica_command.go#L2842)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Oops, reverted.
</details>
Did you still mean to change from `LHS` to `left`? This is now inconsistent with other instances of `"updating LHS descriptor"` in this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[pkg/kv/transport.go, line 186 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8TCo9izwvYCwXaRM:-KUw6T6pShkse9YrzCEs:b-lhl0t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L186)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is local, but can still take appreciable time. is it wise to hold the mutex across this call?
</details>
It can also call back in to the DistSender (for things like splits and intent resolution), leading to deadlocks. This could be the cause of cockroachlabs/production#152.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 1101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUw7C2Dm5JDTA1lKHAm:-KUw7C2Dm5JDTA1lKHAn:b-fxo5tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/dist_sender.go#L1101)):*
> ```Go
> 
> 			// Move the new lease holder to the head of the queue for the next retry.
> 			transport.MoveToFront(leaseHolder)
> ```

Why change this now? I share @petermattis's apprehension about this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 1055 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs5Igkdx_36j1BduXl:-KUwlGa9z8O_fZE85T7U:b-qfy1xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/dist_sender.go#L1055)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> where did you read that this is the error code to check for? aren't there other codes which correspond to the definitely-not-sent condition?
</details>
I looked at the GRPC source code. This is the fail fast condition and it's an error code which isn't used for any other purpose.

---

*[pkg/kv/dist_sender.go, line 1074 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs7jmmiQcadctcl0hV:-KUwlq5dWpo-6rVLBFNB:b-7gk0df) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/dist_sender.go#L1074)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> shouldn't this be `log.ErrEvent(opts.ctx, err.Error())`?
</details>
That would lose the very-purposeful verbosity level specified here. I'm still not sure what appropriate usage of the recent explosion of new log/tracing/event APIs is. I'm wrapping `log.ErrEvent` in a `log.V(2)` as a solution, but it doesn't necessarily feel right. Did you have something else in mind?

---

*[pkg/kv/dist_sender.go, line 1101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUw7C2Dm5JDTA1lKHAm:-KUwn00xs62-u0LG0e4T:bv86078) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/dist_sender.go#L1101)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why change this now? I share @petermattis's apprehension about this. 
</details>
The unittest actually requires it. The original request goes to local node, redirects to leader where we hold up the response, times out,  [we transfer the lease here], dist sender goes to final node, which returns `NotLeaderError` pointing back to local node, so this particular change allows us to revisit the local node.

---

*[pkg/kv/transport.go, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8D2_tPEtJKcyQm3s:-KUwv4zaTdppBD_310Jh:b-bu91xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L102)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It does not appear to return an error.
</details>
Updated

---

*[pkg/kv/transport.go, line 186 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8TCo9izwvYCwXaRM:-KUx0NJaDdHc7EYWohkK:b-t1h7ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L186)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It can also call back in to the DistSender (for things like splits and intent resolution), leading to deadlocks. This could be the cause of cockroachlabs/production#152.
</details>
Changed locking, but this isn't what was blowing up cockroachlabs/production#152

---

*[pkg/kv/transport.go, line 212 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8ZwT3Yzx9CTQLa6n:-KUx0aBFoukhxxMWBwnK:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L212)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Seems unwise to hold the lock across this line.
</details>
Changed.

---

*[pkg/kv/transport.go, line 216 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8c3uouW21NTPWoce:-KUx0iRSa2XRzxL2lWWP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L216)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> This is repeating the comment on the interface; I suggest removing it altogether.
</details>
Done.

---

*[pkg/kv/transport.go, line 222 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs8nTbH1pbdbeYTwyx:-KUx13dqwdYV3543hKyS:b-ucphv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport.go#L222)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> probably best to replace this with:
> ```
> for i := range gt.orderedClients {
>   client := &gt.orderedClients[i]
> ```
> 
> to avoid the subtlety currently present below.
</details>
I'm just using `gt.orderedClients[i]` everywhere. Clearer.

---

*[pkg/kv/transport_test.go, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9TxbENa__L9WwmLg:-KUx28hQ5eehtg687uv2:b-oq84ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/kv/transport_test.go#L41)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> As discussed in many recent PRs, this is bad practice that results in _this_ line number being reported in failures, which makes it impossible to determine which invocation of this function resulted in an unexpected result. Helpers such as this one should return errors for this reason.
</details>
It would nearly triple the size of this method. And FYI, this is why I made the error a `t.Fatalf`, so that the stack trace shows the line number in the test that resulted in the failure.

---

*[pkg/roachpb/errors.proto, line 179 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs86S2zyxrlM2gULQT:-KUx2QjQQ9OnyvXGMNka:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/roachpb/errors.proto#L179)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "is true"?
</details>
Fixed

---

*[pkg/sql/ambiguous_commit_test.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9uwb8isSX9R7qico:-KUx2WMmEVxa2PDW0a7b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L80)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Fatal?
</details>
Done.

---

*[pkg/sql/ambiguous_commit_test.go, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9s7OfTenR3pRiDG5:-KUx2XBxH4zZApmfWBD0:b-j39u3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> please check these errors.
</details>
It's a `SQLRunner`, which already asserts there are no errors.

---

*[pkg/sql/ambiguous_commit_test.go, line 121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUs9px746AXN4EqGRIY:-KUx2bkqifF8G87Zo2iH:bqrqu5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L121)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what is this magic `3`?
</details>
There are three replicas. Just trying to make sure that whichever replica ends up with the range lease, we choose a connection to a different node for the insert statement. Clarified.

---

*[pkg/sql/ambiguous_commit_test.go, line 131 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUsA9p_ZtV5ZbXfltQw:-KUx3WyB-XTsJQWXdehj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L131)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: use `:=` to avoid overwriting random errors at the outer scope
</details>
Done.

---

*[pkg/sql/ambiguous_commit_test.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUsACvHBYMlSpaWmxIz:-KUx3aifTTDrE8yTUYds:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L140)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: use `:=` to avoid overwriting random errors at the outer scope
</details>
Done.

---

*[pkg/sql/ambiguous_commit_test.go, line 146 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUsA5AkiZBJJHPPm4dq:-KUx3lvAzMqhaBRSN8qa:b-s8cmka) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbd9ded4fd503a7c18252419d092765ad6a5afd/pkg/sql/ambiguous_commit_test.go#L146)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> unchecked error? how did this pass lint?
</details>
Because the `SQLRunner` asserts no error.

---

*[pkg/storage/replica_command.go, line 2842 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154#-KUm1vaM5bO5C4UKHgoe:-KUx4QWToLJffxGcgNbS:b9aezoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b67bbc557fc07d251cf98ad8390aaeb3a8adf588/pkg/storage/replica_command.go#L2842)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Did you still mean to change from `LHS` to `left`? This is now inconsistent with other instances of `"updating LHS descriptor"` in this file.
</details>
This is now consistent with `AdminMerge`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10154)*
<!-- Sent from Reviewable.io -->
